pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/scheduler/IScheduler.java

```diff
@@ -22,9 +22,22 @@
 import org.apache.pulsar.functions.proto.Function.Instance;
 
 import java.util.List;
+import java.util.Set;
 
 public interface IScheduler {
 
-    List<Assignment> schedule(List<Instance> unassignedFunctionInstances,
-                              List<Assignment> currentAssignments, List<String> workers);
+    /**
+     * Scheduler schedules assignments to appropriate workers and adds into #resultAssignments
+     * 
+     * @param resultAssignments
```

Comment:
why do we need to pass in "resultAssignment" when we already return that?  Can't you just add the heartbeat assignment to the list of assignments calculated by the scheduler?  It makes the interface confusing

Reply:
> list of assignments calculated by the scheduler?

1. earlier in this PR it was calculated by a RoundRobinScheduler only but then we decided to remove heartbeat logic out from Scheduler-Impl(RoundRobinScheduler) and move it to SchedulerManager as this logic will be same across different Scheduler-implementation.
2. `getUnassignedFunctionInstances(..)` already iterates through all list of instances, current-assignments and prepares list of unassignment instances so, scheduler can schedule it. so, if we just skip it heartbeat function in that call then it requires to iterate to the `allInstance` (`O(n)`)  list again to find out list of heart-beat function again.
So, to optimize it , we are adding heartbeat function to assignment-list in the same iteration and pass the same result-list to the scheduler so, scheduler can add new assignment to the list. I don't think it will confuse the interface because, if caller (SchedulerManager) is modifying and sharing the result-assignment-list then it can be passed to the implementation-scheduler where they can add more assignments.

Reply:
So when you iterate through the instances, why can't you just put the heart beat assignments in a separate list and then merge the assignment lists at the end after the scheduler calculates actual assignments.  

The interface is confusing  because you are returning the same list that is also a parameter to the "schedule" method which is redundant.  If you are passing in resultAssignment and adding new assignment to the list, you don't even need to return it.  However, I would suggest what I mentioned above and use separate lists and merge them at the end.

Reply:
> why can't you just put the heart beat assignments in a separate list 

yes, i wanted to avoid creating garbage (temp maps/lists) if possible at every place to avoid gc-issues in long run. and if SchedulerManager is anyway modifying assignments then I don't see any issue for passing the same list to the sub-sequent chain-processor.

Reply:
why would there be gc issues? All that needs to be done:

resultAssignments.addAll(assignments)

Since we are in java, only pointers are added to the list.  Yes there might be some extra "metadata" related to the arraylist created in the scheduled that would need to gc ed but that would be negligible.  We are not going to be allocating much more memory